### ISSUE TITLE
Fix S3353 FP: Variable is used within an Expression

### ISSUE THREAD

**Issue by lg2de** [SonarSource/sonar-dotnet#4015]
### Description

S3353 is reported for variable if it is used within an `Expression`.
Following the suggestion to make the variable `const` will change the expression.

If the expression is analyzed in frameworks like "NHibernate" the changed expression cause totally different behavior.

I think the rule should not apply for variables using within any `Expression`.

### Repro steps

```cs
namespace SonarDotnetS3353
{
    public class Class1
    {
        public void Method()
        {
            string localVariable = null;

            this.ExpressionMethod(() => localVariable);
        }

        private void ExpressionMethod(Expression<Func<string>> expression)
        {
            if (expression.Body is MemberExpression)
            {
                return;
            }

            throw new InvalidOperationException(
                $"The expression body is a {expression.Body.GetType().Name}, but FieldExpression is expected.");
        }
    }
}
```

When changing `localVariable` to `const string` the analyzing method throws an exception.

### Expected behavior

Rule S3353 should not apply.

### Actual behavior

`SonarDotnetS3353\Class1.cs(10,20,10,41): warning S3353: Add the 'const' modifier to 'localVariable'.`

### Known workarounds

N/A

### Related information

* C#/VB.NET Plugins version: 8.17.0.26580
* Visual Studio version: 2019 16.8.3
* dotnetcore version: 3.1
* Operating System: Windows

**Comment by pavel-mikula-sonarsource** [comment]
Hi @lg2de ,

Thank you for reporting this interesting case. I confirm it as a False Positive.

**Comment by lg2de** [comment]
Maybe the next days I'll try to solve this with a PR.

**Comment by lg2de** [comment]
I failed to fix it. The variations of `Expression` I faced are too complex (for me).

There seams already to be a check for variable usage within `Expression`, but it is limited to assignment.
I'm not sure whether every usage within an expression should stop raising S3533 or not.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @lg2de, thank you for giving it a try.
